### ISSUE TITLE
Add SMM Base HOB for SmBase relocation (Bugzilla Bug 4337)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9323]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 4337
Date: 2023-02-07T21:44:54+00:00
From: @jiaxinwu
To:   @jiaxinwu
CC:   @lgao4, @jiaxinwu

Last updated: 2023-05-06T02:44:53+00:00

**Comment by tianocore-issues** [comment]
## Comment 20999
Date: 2023-02-07 21:44:54 +0000
From: @jiaxinwu

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): @jiaxinwu

Background:

Existing SMBASE Relocation is in the PiSmmCpuDxeSmm driver, which will relocate the SMBASE of each processor by setting the SMBASE field in the saved state map (at offset 7EF8h) to a new value. The RSM instruction reloads the internal SMBASE register with the value in SMBASE field when each time it exits SMM. All subsequent SMI requests will use the new SMBASE to find the starting address for the SMI handler (at SMBASE + 8000h). 

Due to the default SMBASE for all x86 processors is 0x30000, the APs&apos; 1st SMI for rebase has to be executed one by one to avoid the CPUs over-writing each other&apos;s SMM Save State Area (see existing SmmRelocateBases() function), which means the next AP has to wait for the previous AP to finish its 1st SMI, then it can call into its 1st SMI for rebase via Smi Ipi command, thus leading the existing SMBASE Relocation has to be running in series. Besides, it needs very complex code to handle the AP exit semaphore (mRebased[Index]), which will hook return address of SMM Save State so that semaphore code can be executed immediately after AP exits SMM for SMBASE relocation (see existing SemaphoreHook() function). 

This is the request to add the new SMM Base HOB for any PEI module to do the SmBase relocation ahead of PiSmmCpuDxeSmm driver and store the relocated SmBase address in array for each Processors. When the SMBASE relocation happens in a PEI module, the PEI module shall produce the SMM_BASE_HOB in HOB database which tells the PiSmmCpuDxeSmm driver (runs at a later phase) about the new SMBASE for each CPU thread. PiSmmCpuDxeSmm driver installs the SMI handler at the SMM_BASE_HOB.SmBase[Index]+0x8000 for CPU thread Index. When the HOB doesn&apos;t exist, PiSmmCpuDxeSmm driver shall relocate and program the new SMBASE itself (keep existing SMBASE Relocation way).

With above SMM Base Hob support, PiSmmCpuDxeSmm does not need the RSM instruction to do the SMBASE Relocation. SMBASE Register for each processors have already been programmed and all SMBASE address have recorded in SMM Base Hob. So the same default SMBASE Address (0x30000) will not be used, thus the CPUs over-writing each other&apos;s SMM Save State Area will not happen in PiSmmCpuDxeSmm driver. This way makes the first SMI init can be executed in parallel and save boot time on multi-core system. Besides, Semaphore Hook code logic is also not required, which will greatly simplify the SMBASE Relocation flow.

Note:
This is the new way that firmware can program the SMBASE independently of the RSM instruction. The PEI code performing this logic will not be open sourced, similarly to other things that are kept binary-only in the FSP. Due to the register difference in different vender, and it has not been documented in the Intel SDM yet, we need a new binary-only interface for SMM Base HOB.
